src/mqtt_client.rs

```diff
@@ -416,6 +443,12 @@ where
                     };
                 }
 
+                // Replay QoS 1 messages
```

Comment:
I don't think we should be sending QoS level 1 messages here, as we don't know if the connection will be accepted.

That being said, until we incorporate some kind of timer, I don't know where pending-ACK messages _should_ get resent. Perhaps we should add a `resend()` API to allow applications to explicitly request resend transmission when there's pending messages? Or we could do the resend whenever someone calls `publish()`

Reply:
If what connection will be accepted? I thought that in this branch we've already received CONNACK and this is after all the processing of the message so I assumed we can do this here.

As for why I did this, MQTT 5 - 4.4
```
When a Client reconnects with Clean Start set to 0 and a session is present, both the Client and Server MUST resend any unacknowledged PUBLISH packets (where QoS > 0) and PUBREL packets using their original Packet Identifiers. This is the only circumstance where a Client or Server is REQUIRED to resend messages. Clients and Servers MUST NOT resend messages at any other time [MQTT-4.4.0-1]
```

When I think about it this is also the only time it makes sense to retry anything, because if everything works well the Server must send PUBACK and if there is an issue it will surely disconnect and then we retry on new connection.

I'm not against doing the `resend` API, but I'm not sure if there would be any advantage to it. Well maybe in case one of the messages is malformed and then that would cause disconnect during the replay before the user was able to do anything, but then they would not be able to resend manually either.

Reply:
After reading through the MQTT specification, you're correct that this is the correct time for sending them (immediately after a reconnect ConnAck is received).

My main concern is that many network stacks have an internal buffer that may fill up, so we won't necessarily be able to send _all_ pending messages (e.g. might be able to send first 3, but not any more). We need to have a flag for which messages have been retransmitted in the current session so that we can handle sending them in the `poll()` loop some time in the future (once the network stack has transmitted data).

Reply:
I can't find anything except checking the return value of `send` but we cannot really stop mid-packet. Maybe we could keep some kind of state but that would have to be not just which packets were resent, but also an offset of last packet in-flight. And we would have to check before every publish that there is no packet. Otherwise we would have to block there until data at least until packet boundary can be buffered.

Otherwise we could expose some kind of `resend` API you mentioned earlier and mandate that the user has to keep calling it after reconnecting (or are reconnects hidden from users?) until it returns a value indicating success. This would allow the process to not block but this would be somewhat more difficult to use.

Reply:
This seems complicated enough that we should just open an issue for it and fix it later. There's enough in this PR as-is. I'll open up a tracking issue and post it here.